core/services/pipeline/scheduler.go

```diff
@@ -44,8 +73,60 @@ func newScheduler(ctx context.Context, p *Pipeline, vars Vars) *scheduler {
 		resultCh: make(chan TaskRunResult),
 	}
 
-	for _, task := range roots {
-		run := &memoryTaskRun{task: task, vars: s.vars.Copy()}
+	// if there's results already present on Run, then this is a resumption. Loop over them and fill results table
+	for _, r := range run.PipelineTaskRuns {
+		result := Result{}
+
+		task := p.ByDotID(r.DotID)
+
+		if task == nil {
+			panic("can't find task by dot id")
+		}
+
+		if r.IsPending() {
+			continue
+		}
+
+		if r.Error.Valid {
+			result.Error = errors.New(r.Error.String)
+		}
+
+		if !r.Output.Null {
+			result.Value = r.Output.Val
+		}
+
+		s.results[task.ID()] = TaskRunResult{
+			Task:   task,
+			Result: result,
+		}
+
+		// store the result in vars
+		if result.Error != nil {
+			s.vars.Set(task.DotID(), result.Error)
+		} else {
+			s.vars.Set(task.DotID(), result.Value)
+		}
+
+		// mark all outputs as complete
+		for _, output := range task.Outputs() {
+			id := output.ID()
+			s.dependencies[id]--
+		}
+	}
+
+	// immediately schedule all doable tasks
+	for id, task := range p.Tasks {
+		// skip tasks that are not ready
+		if s.dependencies[id] != 0 {
+			continue
+		}
+
+		// skip finished tasks
+		if _, ok := s.results[id]; ok {
```

Comment:
tiny nit: can we call it `exists` instead of `ok`?

Reply:
I've always seen go stdlib use `ok` for these, but I'll rename